### PR TITLE
Add a feature to lookup a file, and get it.

### PR DESCRIPTION
**Pull request by rdemongeot**
##### SUMMARY
Need to store files into BitWarden / ValtWarden; and get them from Ansible.

### CODE REVIEW

File: plugins/lookup/bitwarden.py [hunk 1]
```diff
@@ -31,8 +31,17 @@
...
+      - Field to fetch. Leave unset to fetch whole response.
+      - Mutually exclusive with O(attachment).
+    type: str
+  attachment:
+    description:
+      - Name of the attachment to download from the item.
+      - When set, the plugin will download the attachment content in raw format.
+      - Mutually exclusive with O(field).
     type: str
+    version_added: 11.0.0
```
Comment (russoz) [inline plugins/lookup/bitwarden.py]:
```suggestion
    version_added: 12.0.0
```

File: plugins/lookup/bitwarden.py [hunk 2]
```diff
@@ -72,6 +81,22 @@
...
+- name: "Get attachment 'vpn-server.key' from Bitwarden record named 'VPN Config'"
+  ansible.builtin.debug:
+    msg: >-
+      {{ lookup('community.general.bitwarden', 'VPN Config', attachment='vpn-server.key') }}
+
+- name: "Save attachment to file"
+  ansible.builtin.copy:
+    content: "{{ lookup('community.general.bitwarden', 'VPN Config', attachment='vpn-server.key') | first }}"
+    dest: /etc/vpn/server.key
+    mode: '0600'
```
Comment (russoz) [inline plugins/lookup/bitwarden.py:93]: Not necessarily a problem, but I can't help remembering that the lookup plugins run on the controller, whilst the `copy` task (unless otherwise set) will run on the target(s). This feature has potential to make a dent on the network bandwidth, depending on the file size and number of targets. It would be nice to have at least a warning in the documentation, for new and possibly unwise users.

That begin said, when `slurp`ing a file, the result comes encoded as base64 - I wonder why that is so and whether it would make sense to make that in this case as well - take this with a large grain of salt, I have no hard evidence to support any claim.

File: plugins/lookup/bitwarden.py (hunk 1 above)
Comment (rdemsystems) [reply plugins/lookup/bitwarden.py]: Ok, updated.

File: plugins/lookup/bitwarden.py (hunk 2 above)
Comment (rdemsystems) [reply plugins/lookup/bitwarden.py:93]: A comment was putted just now.

Don't sure that store file from bitwarden is the better idea; but works for now, with couple of stored files.

For network traffic, from Ansible computer to target, any copy will generate network traffic. In this case we'll add from bitwarden/vaultwarden to ansible computer.

**Comment by rdemsystems** [comment]
> Hi @rdemongeot # # Thank you for the contribution! Please add a [changelog fragment]

Can you verify if i'm correct or not?

Regards,

File: changelogs/fragments/10917-bitwarden-attachement.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - bitwarden - add availbility to get Attachment file from bitwarden (https://github.com/ansible-collections/community.general/pull/10917).
```
Comment (felixfontein) [inline changelogs/fragments/10917-bitwarden-attachement.yml]:
```suggestion
  - bitwarden lookup plugin - add availbility to get attachment file from bitwarden (https://github.com/ansible-collections/community.general/pull/10917).
```

Comment (rdemsystems) [reply changelogs/fragments/10917-bitwarden-attachement.yml]: updated.

File: changelogs/fragments/10917-bitwarden-attachement.yml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - bitwarden lookup plugin - add availbility to get attachment file from bitwarden (https://github.com/ansible-collections/community.general/pull/10917).
```
Comment (felixfontein) [inline changelogs/fragments/10917-bitwarden-attachement.yml:2]: Do you mean "ability" or "availability" (an "a" is missing for that)? I think the former makes more sense:
```suggestion
  - bitwarden lookup plugin - add ability to get attachment file from bitwarden (https://github.com/ansible-collections/community.general/pull/10917).
```

File: plugins/lookup/bitwarden.py [hunk 5]
```diff
@@ -31,8 +31,17 @@
...
   field:
-    description: Field to fetch. Leave unset to fetch whole response.
+    description:
+      - Field to fetch. Leave unset to fetch whole response.
+      - Mutually exclusive with O(attachment).
+    type: str
+  attachment:
+    description:
+      - Name of the attachment to download from the item.
+      - When set, the plugin will download the attachment content in raw format.
```
Comment (felixfontein) [inline plugins/lookup/bitwarden.py:41]: What does "raw format" mean? Do you mean "without further processing"? What if the file is binary data (Ansible doesn't like to handle binary data, one usually encodes it as Base64 because of that)?

File: plugins/lookup/bitwarden.py [hunk 6]
```diff
@@ -72,6 +81,25 @@
...
       {{ lookup('community.general.bitwarden', 'bafba515-af11-47e6-abe3-af1200cd18b2', search='id', field='password') | first }}
 
+- name: "Get attachment 'vpn-server.key' from Bitwarden record named 'VPN Config'"
+  ansible.builtin.debug:
+    msg: >-
+      {{ lookup('community.general.bitwarden', 'VPN Config', attachment='vpn-server.key') }}
+
+- name: "Save attachment to file"
+  ansible.builtin.copy:
+    content: "{{ lookup('community.general.bitwarden', 'VPN Config', attachment='vpn-server.key') | first }}"
```
Comment (felixfontein) [inline plugins/lookup/bitwarden.py:91]: While this isn't the first time `| first` was used in the examples: what's the goal of using `| first` here? Does this always return a list of more than one elemnts? If yes, `| first` is fine. If not, `| first` is problematic because `lookup()` itself is problematic - if the lookup plugin returns a one-element list, it automatically takes the single element and returns that instead of the list. (And I think you get `None` or `""` if the list is empty - forgot which one.) Usually it's better to use `query()`, since that will always return the list, even if it has zero or one elements.

File: plugins/lookup/bitwarden.py [hunk 7]
```diff
@@ -232,6 +261,33 @@ def get_field(self, field, search_value, search_field="name", collection_id=None
...
+
+        attachment_contents = []
+        for match in matches:
+            item_id = match.get('id')
+            if not item_id:
+                raise AnsibleError(f"Item {match.get('name', 'unknown')} has no ID")
+
+            try:
+                params = ['get', 'attachment', attachment_name, '--itemid', item_id, '--raw']
+                out, err = self._run(params)
```
Comment (felixfontein) [inline plugins/lookup/bitwarden.py:281]: `out` and `err` are unicode strings; the output of the CLI is parsed as UTF-8. Is this what is meant by "raw format"?

(Can Bitwarden actually store binary files, or must all files be UTF-8 encoded text?)

File: changelogs/fragments/10917-bitwarden-attachement.yml (hunk 4 above)
Comment (russoz) [reply changelogs/fragments/10917-bitwarden-attachement.yml:2]: It is a bit odd as "availability", indeed.

**Comment by felixfontein** [comment]
Please note that this PR now has a conflict since we reformatted all code in the collection (see #10999).

**Comment by felixfontein** [comment]
Ping @rdemongeot 

needs_info

File: plugins/lookup/bitwarden.py (hunk 5 above)
Comment (rdemongeot) [reply plugins/lookup/bitwarden.py:41]: It's the name in VaultWarden.

Comment (rdemongeot) [reply plugins/lookup/bitwarden.py:41]: and i havn't test any edge cases with malformed data to read from vaultwarden

File: plugins/lookup/bitwarden.py (hunk 6 above)
Comment (rdemongeot) [reply plugins/lookup/bitwarden.py:91]: Don't remember all the subjects, but i'm not a develloper, so it was trying to copy/paste from similar functions, and adapt them.

If i remember fine,  bitwarden will effectively answer a list of 0 to N files

File: plugins/lookup/bitwarden.py (hunk 7 above)
Comment (rdemongeot) [reply plugins/lookup/bitwarden.py:281]: Bitwarden can store binary files.

**Comment by rdemongeot** [comment]
> Ping @rdemongeot
> 
> needs_info

Sorry for the delay.

For now, i havn't time / need to re-work on this feature. If it will be at risk, just drop this PR.

**Comment by felixfontein** [comment]
Then let's close it for now. Feel free to create a new PR when you're ready to work on this again.

Thanks for your work on this!
